core/src/main/java/com/linecorp/armeria/client/HAProxyHandler.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import static io.netty.handler.codec.haproxy.HAProxyCommand.PROXY;
+import static io.netty.handler.codec.haproxy.HAProxyProtocolVersion.V2;
+import static io.netty.handler.codec.haproxy.HAProxyProxiedProtocol.TCP4;
+
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.client.proxy.HAProxyConfig;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.haproxy.HAProxyMessage;
+import io.netty.handler.codec.haproxy.HAProxyMessageEncoder;
+import io.netty.handler.proxy.ProxyConnectException;
+
+class HAProxyHandler extends ChannelOutboundHandlerAdapter {
+    private static final Logger logger = LoggerFactory.getLogger(HAProxyHandler.class);
+
+    private final HAProxyConfig haProxyConfig;
+
+    HAProxyHandler(HAProxyConfig haProxyConfig) {
+        this.haProxyConfig = haProxyConfig;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        ctx.pipeline().addBefore(ctx.name(), null, HAProxyMessageEncoder.INSTANCE);
+        super.handlerAdded(ctx);
+    }
+
+    @Override
+    public void connect(ChannelHandlerContext ctx, SocketAddress remoteAddress,
+                        SocketAddress localAddress, ChannelPromise promise) throws Exception {
+        promise.addListener(f -> {
+            if (!f.isSuccess()) {
+                return;
+            }
+            try {
+                ctx.write(createMessage(haProxyConfig, ctx.channel())).addListener(f0 -> {
+                    if (f0.isSuccess()) {
+                        ctx.pipeline().remove(HAProxyMessageEncoder.INSTANCE);
+                    } else {
+                        ctx.fireExceptionCaught(new ProxyConnectException(f0.cause()));
+                        ctx.close();
+                    }
+                });
+            } catch (Exception e) {
+                ctx.channel().eventLoop().execute(() -> {
```

Comment:
Can we follow the pattern?
```java
final EventLoop eventLoop = ctx.channel().eventLoop();
if (eventLoop.inEventLoop()) {
    ctx.pipeline().fireUserEventTriggered(new ProxyConnectException(e));
    ctx.close();
} else {
    ....
}
```

Reply:
Actually this is already in the event loop -- I am just rescheduling since other handlers aren't added yet.

Reply:
> other handlers aren't added yet.

I think I don't fully understand the flow of how handlers are added.
Could you elaborate more on this, please? 😄

Reply:
I don't understand fully either 😄 , but here is my understanding:

1. `HAProxyHandler` is the added as the first handler and the listener is added immediately in the same thread when invoking `HttpChannelPool.bootstrap.connect`.
2. Afterwards, the listener is added at `HttpChannelPool` for adding `HttpSessionHandler` (same thread as 1)
https://github.com/line/armeria/blob/d30e82def4d3229302f4a9306bcb9b51c00b4dcc/core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java#L405
3. When `connect` is successful and an exception is thrown, the `HAProxyHandler` listener is invoked first. At this time, `HttpChannelPool` listener isn't invoked yet and the `HttpSessionHandler` isn't in the pipeline. Consequently, `userEventTriggered` isn't handled by `HttpSessionHandler`.

Although `HttpClientPipelineConfigurator` doesn't add `listeners` directly, I was assuming the following block is aiming for the same effect as well.
https://github.com/line/armeria/blob/d30e82def4d3229302f4a9306bcb9b51c00b4dcc/core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java#L348

Reply:
Ah, thanks a lot for the detailed explanation. 🙇‍♂️ 
Could you add a comment for this?
`// Call fireUserEventTriggered in the executor to execute it after the HttpSessionHandler is added to the pipeline.`